File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3035_maximum-palindromes-after-operations/solutions/c/solution.c
#include <stdbool.h>
#include <string.h>

int maxPalindromesAfterOperations(char** words, int wordsSize) {
    int charCount[26] = {0};
    
    // Count frequency of each character across all words
    for (int i = 0; i < wordsSize; i++) {
        for (int j = 0; words[i][j]; j++) {
            charCount[words[i][j] - 'a']++;
        }
    }
    
    int totalPairs = 0;
    int oddCount = 0;
    
    // Calculate total pairs and count of odd characters
    for (int i = 0; i < 26; i++) {
        totalPairs += charCount[i] / 2;
        if (charCount[i] % 2 == 1) {
            oddCount++;
        }
    }
    
    // Determine maximum number of palindromes
    int maxPalindromes = totalPairs / (wordsSize / 2);
    if (wordsSize % 2 == 1) maxPalindromes++;
    
    return (maxPalindromes > wordsSize) ? wordsSize : maxPalindromes;
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3035_maximum-palindromes-after-operations/solutions/java/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maxPalindromesAfterOperations(String[] words) {
        Map<Character, Integer> frequency = new HashMap<>();
        
        // Count all character frequencies across all words
        for (String word : words) {
            for (char c : word.toCharArray()) {
                frequency.put(c, frequency.getOrDefault(c, 0) + 1);
            }
        }
        
        int maxPalindromes = 0;
        
        // Determine maximum number of palindromes that can be formed
        int oddCount = 0;
        for (int count : frequency.values()) {
            if (count % 2 != 0) {
                oddCount++;
            }
        }
        
        // Each palindrome needs at most one odd character count
        // The number of palindromes that can be formed is the total number of strings minus 
        // the surplus of odd character counts (more odd counts than words would mean we can't form that many palindromes)
        maxPalindromes = words.length - Math.max(0, oddCount - words.length);
        
        return maxPalindromes;
    }
}
